RavenDB-19608 Add AggressivelyCacheForAync to avoid sync usage in asy…

ayende opened the pull request:
…nc manner

### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19608 

### Additional description

Allow to setup aggressive caching in a way that works nicely with async code.

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

ayende committed:
RavenDB-19608 Add AggressivelyCacheForAync to avoid sync usage in async manner

ppekrol commented:
It caused a lot of test failures, at glance, not sure why, do not see anything suspicious here

ppekrol committed:
RavenDB-19608
- added missing AggressivelyCacheForAsync overload
- cannot use AsyncHelpers for AggressivelyCacheFor because the AsyncLocal looses context

ayende commented:
Any idea why? My goal was to reduce duplication, can't see the difference in behavior

ppekrol committed:
RavenDB-19608 fixed issue with AsyncLocal usage in AggresiveCache when async/await is used

ppekrol committed:
RavenDB-19608 renames

danielmarbach commented on src/Raven.Client/Documents/DocumentStore.cs:
```diff
@@ -344,24 +345,74 @@ public override IDisposable AggressivelyCacheFor(TimeSpan cacheDuration, string
+        {
+            var release = SetAggressiveCache(cacheDuration, mode, ref database);
+
+            return AsyncHelpers.RunSync(() => FinalizeAggressiveCacheAsync(release, mode, database));
```
Is there a reason why the helper doesn't provide a state-based overload that would avoid the closure?

danielmarbach commented on src/Raven.Client/Documents/DocumentStore.cs:
```diff
@@ -344,24 +345,74 @@ public override IDisposable AggressivelyCacheFor(TimeSpan cacheDuration, string
+            requestExecutor.AggressiveCaching.Value = newValue;
 
-            re.AggressiveCaching.Value = @new;
+            return new DisposableAction(() => requestExecutor.AggressiveCaching.Value = oldValue);
```
Is there a reason why the action doesn't provide a state-based overload that would avoid the closure?

danielmarbach commented on src/Raven.Client/Documents/DocumentStore.cs:
```diff
@@ -344,24 +345,74 @@ public override IDisposable AggressivelyCacheFor(TimeSpan cacheDuration, string
+        /// we provide is current or not, but will serve the information directly from the local cache
+        /// without touching the server.
+        /// </remarks>
+        public override Task<IDisposable> AggressivelyCacheForAsync(TimeSpan cacheDuration, AggressiveCacheMode mode, string database = null)
```
Why doesn't this use `ValueTask`? It seems the majority case is that you are already connected and then you could entirely avoid the task allocation

danielmarbach commented on test/SlowTests/Issues/RavenDB_16358.cs:
```diff
@@ -57,10 +57,10 @@ public async Task AggressiveCacheWithTimeoutTestAsync()
         {
             using (var session = store.OpenAsyncSession())
-            using (session.Advanced.DocumentStore.AggressivelyCacheFor(TimeSpan.FromMilliseconds(300)))
+            using (await session.Advanced.DocumentStore.AggressivelyCacheForAsync(TimeSpan.FromMilliseconds(300)))
```
By the way this is a classical pitfall of returning `Task<IDisposable>` and I bet this will lead to many support hazzles.

ppekrol commented:
Hi @danielmarbach 

I've addressed some of your comments here: https://github.com/ravendb/ravendb/pull/15487

Thanks for feedback!
